Implement label-based container discovery for Redis Dev Service

patrox opened the pull request:
The PR started as a copy-paste solution for label-based container discovery for Redis dev service -
in order to handle: https://github.com/quarkusio/quarkus/issues/18029.

Based on these two PRs:

- Kafka: https://github.com/quarkusio/quarkus/pull/17837
- AMQP: https://github.com/quarkusio/quarkus/pull/17980

As many of you have noticed this required refactoring - so for that purpose I introduced two new classes:

- [ContainerLocator](https://github.com/quarkusio/quarkus/pull/18181/files#diff-46fcf1e20a4ba63f591741375828efe01cd293e303ff664663f4262476fb5238)
- [ContainerAddess](https://github.com/quarkusio/quarkus/pull/18181/files#diff-efd6f27826c11155e7d487d185a3eed8e0eaa230aa131ba02f87d1937da0ef06)

Both of them are located in a new module called simply: `Quarkus - DevServices - Common`.
The commits were not squashed (yet) for easier review.

patrox committed:
Implement label-based container lookup for Redis Dev Service

patrox committed:
Introduce a common label-based container locator

Refactor redis-client devservice to use a common container locator
Refactor kafka-client devservice to use a common container locator
Refactor amqp-client devservice to use a common container locator

patrox commented:
@machi1990 @stuartwdouglas @cescoffier @Ladicek Can I please kindly ask you for a review?

cescoffier commented on docs/src/main/asciidoc/redis.adoc:
```diff
@@ -89,6 +89,20 @@ docker run --ulimit memlock=-1:-1 -it --rm=true --memory-swappiness=0 --name red
 If you use <<dev-services,DevServices>>, launching the container manually is not necessary!
 ====
 
+== Shared server
```
Can you extract the dev services for Redis documentation into its own file as we did for Kafka and AMQP.
We are refactoring the documentation to distinguish the "getting started" part from the "reference" part and the configuration of dev services is in the later part.

patrox replied:
Sure, I'll do that.

cescoffier commented on extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:
```diff
@@ -0,0 +1,23 @@
+package io.quarkus.devservices.common;
+
+public class ContainerAddress {
```
Unfortunately, most of the time it won't be enough.
We may need to add "extra" parameters. 

We can do that later.

patrox replied:
Yes, I'm aware of that. The current state of this class is to address the combined requirements for:

- Redis
- Kafka
- AMQP

Future cases might require additional fields.

cescoffier commented on extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:
```diff
@@ -0,0 +1,23 @@
+    private final String ipAddress;
+    private final int port;
+
+    public ContainerAddress(String ipAddress, Integer port) {
```
can the port be `null`? If not, why `Integer` and not `int`?

patrox replied:
Theoretically, it can, as the [ContainerPort#getPublicPort](https://github.com/docker-java/docker-java/blob/master/docker-java-api/src/main/java/com/github/dockerjava/api/model/ContainerPort.java#L67) is annoted with `CheckForNull`.

patrox replied:
Correct, so I added an explicit check for this unlikely event, to be on the safe side.

cescoffier commented on extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:
```diff
@@ -0,0 +1,23 @@
+        this.port = port;
+    }
+
+    public String getIpAddress() {
```
What about just `getHost`.
First, it may not be an IP, but a name (like "localhost"), and from a user POV, we just need the host name (how is it resolved is irrelevant).

patrox replied:
Yes, this is a better name.

cescoffier commented on extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    public String getUrl() {
+        return String.format("%s:%d", ipAddress, port);
```
Should we add an optional scheme? Would that make sense?

patrox replied:
At the moment it wasn't (isn't) needed - I wanted to keep it simple.

machi1990 reviewed (commented):
lgtm, thanks @patrox 

Can you squash the commits?

cescoffier commented on extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java:
```diff
@@ -0,0 +1,61 @@
+            if (containerPort != null) {
+                final ContainerAddress containerAddress = new ContainerAddress(containerPort.getIp(),
+                        containerPort.getPublicPort());
+                log.infof("Dev Services container locator found: %s (%s). "
```
why "locator"?

patrox replied:
This is how the class which encapsulates the logic is called (`Container Locator` - used to locate containers).

Ladicek replied:
Except what you _found_ here is not a _container locator_; it's the _container_.

patrox replied:
> Except what you _found_ here is not a _container locator_; it's the _container_.

Correct, the message will be changed to be more clear.

cescoffier commented on extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java:
```diff
@@ -44,9 +56,9 @@
     @BuildStep(onlyIfNot = IsNormal.class, onlyIf = IsDockerRunningSilent.class)
     public void startRedisContainers(LaunchModeBuildItem launchMode,
-            BuildProducer<RunTimeConfigurationDefaultBuildItem> runTimeConfiguration,
+                                     BuildProducer<RunTimeConfigurationDefaultBuildItem> runTimeConfiguration,
```
Why these spaces?

patrox replied:
I believe that this was autoformatted by my IntelliJ to align it with the other params.
So the `runTimeConfiguration` is aligned with `launchMode` above.

I can revert it as it's just formatting.

Ladicek replied:
`mvn clean compile` should fix that automatically (using the Quarkus formatter style).

patrox replied:
Fixed.

cescoffier commented on extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java:
```diff
@@ -44,9 +56,9 @@
+                                     BuildProducer<RunTimeConfigurationDefaultBuildItem> runTimeConfiguration,
 
-            BuildProducer<DevServicesNativeConfigResultBuildItem> devConfigProducer, RedisBuildTimeConfig config) {
+                                     BuildProducer<DevServicesNativeConfigResultBuildItem> devConfigProducer, RedisBuildTimeConfig config) {
```
Why these spaces?

patrox replied:
Fixed.

cescoffier commented on extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java:
```diff
@@ -139,7 +151,16 @@ private StartResult startContainer(String connectionName, DevServicesConfig devS
                 .asCompatibleSubstituteFor(REDIS_6_ALPINE);
-        FixedPortRedisContainer redisContainer = new FixedPortRedisContainer(dockerImageName, devServicesConfig.port);
+
+        if (devServicesConfig.shared && launchMode == DEVELOPMENT) {
```
Can this be extracted too?
Maybe not easily.

patrox replied:
I'll take a look at this as well, as it would further reduce the duplicated code.

patrox replied:
@cescoffier I see two possibilities how this can be extracted:

- in a smaller (simpler) scope we could extract the conditional logic for label-based container location (lines 155-160)
- in a bit broader scope we could extract the above __and__ also the other possibility, that is a creation of a new `Container` - which is done when the `shared` container is not used. (lines 155-172).

WDYT?

cescoffier replied:
I would go with the first one.

patrox replied:
@cescoffier I pushed the latest changes, generally, the `ContainerLocater` was made more "functional" by using `Optional` instead of bare `null`. This has shortened the code and additionally, it allowed to compose the method calls in a more fluent way.

cescoffier commented on extensions/redis-client/deployment/src/main/java/io/quarkus/redis/client/deployment/DevServicesProcessor.java:
```diff
@@ -169,12 +190,15 @@ public StartResult(String url, Closeable closeable) {
             super(dockerImageName);
             this.fixedExposedPort = fixedExposedPort;
+            if (serviceName != null) {
+                withLabel(DEV_SERVICE_LABEL, serviceName);
```
What about having a container decorator that adds the label. 
At the moment we use a label-based mechanism, but this is likely going to change. 

(can be done in another PR).

patrox commented:
> lgtm, thanks @patrox
> 
> Can you squash the commits?

Sure, they will be squashed.

Ladicek commented:
Before I get to reviewing (which I'd really like to do per commit), what's the obsession with squashing commits? Clearly someone has made an effort to divide the work into smaller chunks that form a logical progression -- why would we want to lose that?

machi1990 commented:
> Before I get to reviewing (which I'd really like to do per commit), what's the obsession with squashing commits? Clearly someone has made an effort to divide the work into smaller chunks that form a logical progression -- why would we want to lose that?

It is not a obsession, but a suggestion. To me this boils down to two:
- Implementing the label based discovery on the Redis extension using the logic from existing extension
- Avoiding duplicates of the label-based discovery logic by extracting common code in a module and applying it through out. 

I am all +1 on having atomic commits, that can easily be cherry-picked and have the build pass.

Ladicek commented:
OK, I just went through the commits and the last one modifies multiple modules, whereas previously, each module was modified in a separate commit (which is very nice IMHO). In which case, squashing all the commits (except maybe the first one) would probably be best indeed. (Or percolate the changes from the last commit to each individual previous commit, though I can see how that's not the best use of one's time :-) )

patrox commented:
> OK, I just went through the commits and the last one modifies multiple modules, whereas previously, each module was modified in a separate commit (which is very nice IMHO). In which case, squashing all the commits (except maybe the first one) would probably be best indeed. (Or percolate the changes from the last commit to each individual previous commit, though I can see how that's not the best use of one's time :-) )

About the squashing process - I usually like to create a PR with separate commits to make the review a bit easier.
But, before merging it, I prefer to squash them and leave just a small number of commits that are related to separate changes.

In this case, I would leave just two commits:

- Implement label-based container lookup for Redis Dev Service
- Introduce a common label-based container locator + refactoring

patrox commented:
@cescoffier @Ladicek @machi1990 I addressed all of your comments, can I please ask you kindly for a 2nd round?

Please note that the commits will be squashed as discussed above - to 2 separate commits.

machi1990 commented on docs/src/main/asciidoc/redis.adoc:
```diff
@@ -615,21 +615,6 @@ ReactiveRedisClient defaultReactiveRedisClient = ReactiveRedisClient.createClien
 ReactiveRedisClient namedReactiveRedisClient = ReactiveRedisClient.createClient("second");
 ----
 
-
-[[dev-services]]
-=== DevServices (Configuration Free Redis)
-
-Quarkus supports a feature called DevServices that allows you to create various datasources without any config.
-What that means practically, is that if you have docker running and have not configured `quarkus.redis.hosts`,
-Quarkus will automatically start a Redis container when running tests or dev-mode, and automatically configure the connection.
-
-Available properties to customize the Redis DevService.
-include::{generated-dir}/config/quarkus-redis-client-config-group-dev-services-config.adoc[opts=optional, leveloffset=+1]
-
-When running the production version of the application, the Redis connection need to be configured as normal,
-so if you want to include a production database config in your `application.properties` and continue to use DevServices
-we recommend that you use the `%prod.` profile to define your Redis settings.
-
```
Thanks for extracting this into a separate. 

This was previously linked in the beginning of the doc using the "dev-services" anchor. Do you mind updating the two `NOTE` to link to the added doc?

patrox replied:
Sure, I'll do that.

patrox replied:
@machi1990 Done.

Ladicek commented on extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerAddress.java:
```diff
@@ -1,23 +1,27 @@
-        this.ipAddress = ipAddress;
+    public ContainerAddress(String host, Integer port) {
+        Objects.requireNonNull(host, "Host shouldn't be null");
+        Objects.requireNonNull(port, "Port shouldn't be null");
```
Maybe that's just me, but the best signalization that `null` is forbidden would be to not accept `Integer`, but simply `int`.

patrox replied:
Yes, this is also an option - but this will require a check before passing the value to the ctor.
I can refactor it - it's not a problem.

Ladicek commented on docs/src/main/asciidoc/redis-dev-services.adoc:
```diff
@@ -0,0 +1,39 @@
+:extension-status: preview
+include::./attributes.adoc[]
+
+=== DevServices (Configuration Free Redis)
```
Any reason to not call this "Dev Services for Redis"? That's how similar guides are called for Kafka, AMQP or Apicurio Registry.

patrox replied:
@Ladicek Actually this is a result of refactoring, as it's already called `Dev Service for Redis` - as this is how I named it (to be in line with i.e. Kafka guide).

The line which you are referring to is a left-over, so I'll remove it as it doesn't seem to be needed:

![image](https://user-images.githubusercontent.com/476392/123759443-30dc5a80-d8c0-11eb-8875-2e08962dfb5d.png)

cescoffier reviewed (changes requested):
A minor comment still missing (some tiny refactoring), the rest looks great.

patrox commented:
> A minor comment still missing (some tiny refactoring), the rest looks great.

@cescoffier I have this change already prepared locally, I'll push it a bit later.

stuartwdouglas commented on docs/src/main/asciidoc/redis-dev-services.adoc:
```diff
@@ -0,0 +1,37 @@
+:extension-status: preview
+include::./attributes.adoc[]
+
+Quarkus supports a feature called DevServices that allows you to create various datasources without any config.
```
Shouldn't this just go in the main redis doc as a subsection?

cescoffier replied:
Kafka and AMQP already use this structure. The getting started contains admonitions referring to this file. Dev service configuration is more reference content than getting started content.

cescoffier commented on extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java:
```diff
@@ -0,0 +1,58 @@
+        return Arrays.stream(container.getPorts())
+                .filter(containerPort -> containerPort.getPrivatePort() != null)
+                .filter(containerPort -> containerPort.getPublicPort() != null)
+                .filter(containerPort -> containerPort.getPrivatePort() == port)
```
Minor: The 3 filters can be merged in a single one.

patrox replied:
@cescoffier Yes, I did consider it, but I didn't want to make this line too long - as right now it's easier to read.
Actually now, when I saw this code again with a fresh head - I think it could be extracted to a predicate, so we will a single filter - which won't be too long at the same time.

gsmet reviewed (dismissed):
Once the dust has settled, please try to squash the commits a bit to have proper semantic ones if possible.

patrox commented:
> Once the dust has settled, please try to squash the commits a bit to have proper semantic ones if possible.

@gsmet  The commits were squashed and the branch was rebased to the latest ~~`master`~~, sorry `main`.

patrox commented:
@gsmet @Ladicek I have squashed the commits so there are only two commits.
Can you please have a look and approve if there are no more comments from your side ?

Ladicek reviewed (approved):
I think this is OK, though there's certain school of thought (relatively prevalent on the Quarkus team) that prefers anonymous classes over lambdas because of memory footprint. Since the changes from anonymous classes to lambdas in this PR are in `deployment` modules and hence don't affect production builds, I think that should be fine :-)

patrox commented:
> I think this is OK, though there's certain school of thought (relatively prevalent on the Quarkus team) that prefers anonymous classes over lambdas because of memory footprint. Since the changes from anonymous classes to lambdas in this PR are in `deployment` modules and hence don't affect production builds, I think that should be fine :-)

Thanks, @Ladicek for this insightful answer - I was actually wondering what's the rationale for such heavy usage of anonymous classes - and now I know.

machi1990 commented:
Thanks @patrox for the great work.
